Ensemble learning = combining multiple models to make a single model with accuracy greater than the accuracy of each of the model used in ensemble learning.

Voting ensemble = the base models can be different or same, all models train with same data set independently.

If we built a classification voting ensemble, the output would be the mode of the set of outputs produced by each of the base models.
If we built a regression voting ensemble, the output would be the mean of the set of outputs produced by each of the base models.



The assumptions of the base models in ensemble learning were
1) All models should be independent.
2) Accuracy of each model should be greater or equal to 50%

How multiple models combine to give greater accuracy than each of the base model ?

Let's assume we have 3 base models, each with an accuracy of 0.7. I made a voting ensemble.
Now, think in terms of conditional probabilities,
let m1, m2, m3 be 3 models

P(m1 and m2 and m3) = 0.7*0.7*0.7 = 0.343
p(m1 and m2 and m3') = 0.7*0.7*0.3 = 0.147
p(m1 and m2' and m3) = 0.7*0.3*0.7 = 0.147
p(m1 and m2' and m3') = 0.7*0.3*0.3 = 0.063
p(m1' and m2 and m3) = 0.3*0.7*0.7 = 0.147
p(m1' and m2 and m3') = 0.3*0.7*0.3 = 0.063
p(m1' and m2' and m3) = 0.3*0.3*0.7 = 0.063
p(m1' and m2' and m3') = 0.3*0.3*0.3 = 0.27

The final output is decided based on the majority of outputs. To get a majority, at least 2 models should be correct
P(m1 and m2 and m3) = 0.7*0.7*0.7 = 0.343
p(m1 and m2 and m3') = 0.7*0.7*0.3 = 0.147
p(m1 and m2' and m3) = 0.7*0.3*0.7 = 0.147
p(m1' and m2 and m3) = 0.3*0.7*0.7 = 0.147

0.343+0.147+0.147+0.147 = 0.784 = 78.4% 

Intially we have 70% now we achieved 78.4% using voting ensemble.

Now we see a code demo.

In [1]:
import pandas as pd
import numpy as np
from sklearn import datasets
iris = datasets.load_iris()

In [5]:
iris.feature_names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

In [6]:
iris.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [8]:
iris.data

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [9]:
X=pd.DataFrame(iris.data,columns=iris.feature_names)

In [10]:
X.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [29]:
y=pd.DataFrame(iris.target,columns=['Species'])

In [30]:
y.head()

Species
0        0
1        0
2        0
3        0
4        0

In [31]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score

In [32]:
clf1=LogisticRegression(max_iter=1000)
clf2=KNeighborsClassifier()
clf3=RandomForestClassifier()

In [33]:
estimators=[('lr',clf1),('knn',clf2),('rnn',clf3)]

In [34]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X = scaler.fit_transform(X)
y = y.values.ravel()
estimators = [
    ('LR', clf1),
    ('KNN', clf2),
    ('DT', clf3)
]
for i in estimators:
    scores = cross_val_score(i[1], X, y, cv=10, scoring='accuracy')
    print(i[0], np.round(np.mean(scores), 2))


LR 0.96
KNN 0.95
DT 0.95


In [35]:
from sklearn.ensemble import VotingClassifier

HARD VOTING

In [36]:
vc=VotingClassifier(estimators=estimators)
x=cross_val_score(vc,X,y,cv=10,scoring='accuracy')
print(np.round(np.mean(x),2))

0.96


SOFT VOTING

In [37]:
vc=VotingClassifier(estimators=estimators,voting='soft')
x=cross_val_score(vc,X,y,cv=10,scoring='accuracy')
print(np.round(np.mean(x),2))

0.96


WEIGHTED VOTING

In [38]:
for i in range(1,4):
    for j in range(1,4):
        for k in range(1,4):
            vc=VotingClassifier(estimators=estimators,voting='soft',weights=[i,j,k])
            x=cross_val_score(vc,X,y,cv=10,scoring='accuracy')
            print("for i={},j={},k={}".format(i,j,k),np.round(np.mean(x),2))
            

for i=1,j=1,k=1 0.96
for i=1,j=1,k=2 0.96
for i=1,j=1,k=3 0.96
for i=1,j=2,k=1 0.95
for i=1,j=2,k=2 0.95
for i=1,j=2,k=3 0.96
for i=1,j=3,k=1 0.95
for i=1,j=3,k=2 0.96
for i=1,j=3,k=3 0.96
for i=2,j=1,k=1 0.96
for i=2,j=1,k=2 0.96
for i=2,j=1,k=3 0.96
for i=2,j=2,k=1 0.96
for i=2,j=2,k=2 0.95
for i=2,j=2,k=3 0.96
for i=2,j=3,k=1 0.95
for i=2,j=3,k=2 0.96
for i=2,j=3,k=3 0.95
for i=3,j=1,k=1 0.96
for i=3,j=1,k=2 0.96
for i=3,j=1,k=3 0.96
for i=3,j=2,k=1 0.96
for i=3,j=2,k=2 0.96
for i=3,j=2,k=3 0.96
for i=3,j=3,k=1 0.96
for i=3,j=3,k=2 0.96
for i=3,j=3,k=3 0.95


The maximum accuracy we are getting was 96%

We can use same base models instead of different models

Classifiers of same algo

In [40]:
from sklearn.svm import SVC
from sklearn.datasets import make_classification

In [41]:
X,y=make_classification(n_samples=1000,n_features=20,n_informative=15,n_redundant=5,random_state=2)

In [44]:
svm1=SVC(probability=True,kernel='poly',degree=1)
svm2=SVC(probability=True,kernel='poly',degree=2)
svm3=SVC(probability=True,kernel='poly',degree=3)
svm4=SVC(probability=True,kernel='poly',degree=4)
svm5=SVC(probability=True,kernel='poly',degree=5)

In [45]:
estimators=[('svm1',svm1),('svm2',svm2),('svm3',svm3),('svm4',svm4),('svm5',svm5)]

In [46]:
for i in estimators:
    scores = cross_val_score(i[1], X, y, cv=10, scoring='accuracy')
    print(i[0], np.round(np.mean(scores), 2))

svm1 0.85
svm2 0.85
svm3 0.89
svm4 0.81
svm5 0.86


Here we are getting maximun accuracy 89% lets see what happen if we use all this 5 as base models in voting ensemble

In [49]:
vc1=VotingClassifier(estimators=estimators,voting='soft')
x=cross_val_score(vc1,X,y,cv=10,scoring='accuracy')
print(np.round(np.mean(x),2))

0.93


We are getting 93% 

In [50]:
vc1=VotingClassifier(estimators=estimators,voting='hard')
x=cross_val_score(vc1,X,y,cv=10,scoring='accuracy')
print(np.round(np.mean(x),2))

0.91


In [52]:
for i in range(1,4):
    for j in range(1,4):
        for k in range(1,4):
            for p in range(1,4):
                for q in range(1,4): 
                    vc2=VotingClassifier(estimators=estimators,voting='soft',weights=[i,j,k,p,q])
                    x=cross_val_score(vc2,X,y,cv=10,scoring='accuracy')
                    print("for i={},j={},k={},p={},q={}".format(i,j,k,p,q),np.round(np.mean(x),2))

for i=1,j=1,k=1,p=1,q=1 0.93
for i=1,j=1,k=1,p=1,q=2 0.92
for i=1,j=1,k=1,p=1,q=3 0.92
for i=1,j=1,k=1,p=2,q=1 0.93
for i=1,j=1,k=1,p=2,q=2 0.93
for i=1,j=1,k=1,p=2,q=3 0.92
for i=1,j=1,k=1,p=3,q=1 0.92
for i=1,j=1,k=1,p=3,q=2 0.94
for i=1,j=1,k=1,p=3,q=3 0.93
for i=1,j=1,k=2,p=1,q=1 0.92
for i=1,j=1,k=2,p=1,q=2 0.92
for i=1,j=1,k=2,p=1,q=3 0.92
for i=1,j=1,k=2,p=2,q=1 0.93
for i=1,j=1,k=2,p=2,q=2 0.93
for i=1,j=1,k=2,p=2,q=3 0.93
for i=1,j=1,k=2,p=3,q=1 0.94
for i=1,j=1,k=2,p=3,q=2 0.94
for i=1,j=1,k=2,p=3,q=3 0.93
for i=1,j=1,k=3,p=1,q=1 0.91
for i=1,j=1,k=3,p=1,q=2 0.91
for i=1,j=1,k=3,p=1,q=3 0.91
for i=1,j=1,k=3,p=2,q=1 0.92
for i=1,j=1,k=3,p=2,q=2 0.92
for i=1,j=1,k=3,p=2,q=3 0.92
for i=1,j=1,k=3,p=3,q=1 0.93
for i=1,j=1,k=3,p=3,q=2 0.93
for i=1,j=1,k=3,p=3,q=3 0.93
for i=1,j=2,k=1,p=1,q=1 0.93
for i=1,j=2,k=1,p=1,q=2 0.93
for i=1,j=2,k=1,p=1,q=3 0.93
for i=1,j=2,k=1,p=2,q=1 0.92
for i=1,j=2,k=1,p=2,q=2 0.93
for i=1,j=2,k=1,p=2,q=3 0.93
for i=1,j=2,k=1,p=3,q=1 0.92
for i=1,j=2,k=

There is a slight improvemet now I am getting 94% accuracy